pyacq/core/stream.py

```diff
@@ -523,11 +576,348 @@ def get_array_slice(self, index, length, autoswapaxes = True):
+                self._write(fill_start, fill_stop, self._filler)
+                revert_inds[1] = fill_stop-1
+                
+            self._write(self._write_index + 1 - dsize, self._write_index + 1, data)
+                
+            self._set_read_index(index)
+        except:
+            # If there is a failure writing data, revert read/write pointers
+            self._read_index = revert_inds[0]
+            self._write_index = revert_inds[1]
+            raise
+
+    def _set_write_index(self, i):
+        # what kind of protection do we need here?
+        self._write_index = i
+
+    def _set_read_index(self, i):
+        # what kind of protection do we need here?
+        self._read_index = i
+        
+    def _write(self, start, stop, value):
+        # get starting index
+        bsize = self.shape[0]
+        dsize = stop - start
+        i = start % bsize
+        
+        if self.double:
+            self.buffer[i:i+dsize] = value
+            i += bsize
+        
+        if i + dsize <= self.buffer.shape[0]:
+            self.buffer[i:i+dsize] = value
+        else:
+            n = self.buffer.shape[0]-i
+            self.buffer[i:] = value[:n]
+            self.buffer[:dsize-n] = value[n:]
+        
+    def __getitem__(self, item):
+        #print("GET:", item, "write_ind:", self._write_index, "read_ind:", self._read_index, "shape:", self.shape, "buf_shape:", self.buffer.shape)
+        bsize = self.shape[0]
+        
+        if isinstance(item, tuple):
+            first = item[0]
+            rest = (slice(None),) + item[1:]
+        else:
+            first = item
+            rest = None
+        
+        if isinstance(first, int):
+            index = self._interpret_index(first)
+            #print("  index:", first, index, index%bsize)
+            data = self.buffer[index % bsize]
+        elif isinstance(first, slice):
+            start, stop, step = self._interpret_index(first)
+            if self.double:
+                start_ind = start % bsize
+                stop_ind = start_ind + (stop - start)
+                #print("  double:", (start, stop), (start_ind, stop_ind), bsize)
+                data = self.buffer[start_ind:stop_ind]
+            else:
+                break_index = (self._write_index + 1) - ((self._write_index + 1) % bsize)
+                if (start < break_index) == (stop <= break_index):
+                    start_ind = start % bsize
+                    stop_ind = start_ind + (stop - start)
+                    #print("  single; contig:", (start, stop), (start_ind, stop_ind), bsize, break_index)
+                    data = self.buffer[start_ind:stop_ind]
+                else:
+                    #print("  single; discontig:", (start, stop), (start%bsize, stop%bsize), bsize, break_index)
+                    # need to reconstruct from two pieces
+                    data = np.empty((stop-start,) + self.shape[1:], self.buffer.dtype)
```

Comment:
This case is important.
Maybe we could have 2 behaviors for that case:
   * one doing a memcopy at the discontinuity : the case you implement.
   * one without memcopy that always return data1, data2. data2 is None in the normal case data is not None at the discontinuity. This more annoying for the reader but this avoid copy.

example
```
ringbuffer = RingBuffer(double=False, discontinuitymode='nocopy')
data1, data2 = ringbuffer[ind1:ind2]
process(data1)
if data2 is not None:
  process(data2)


```